sat_vapor_pres/sat_vapor_pres_k.F90

```diff
@@ -1730,290 +3533,739 @@ end subroutine lookup_es2_k_0d
+             nbad = nbad+1
+           else
+             del = tmp-dtres*real(ind)
+             esat(i) = TABLE3(ind+1) + del*(DTABLE3(ind+1)+del*D2TABLE3(ind+1))
+             desat(i) = DTABLE3(ind+1) + 2.*del*D2TABLE3(ind+1)
+           endif
+         enddo
+       end select
+     end select
+   end select
 
  end subroutine lookup_es3_des3_k_1d
 
 !#######################################################################
 
  subroutine lookup_es3_des3_k_0d (temp, esat, desat, nbad)
- real, intent(in)     :: temp
- real, intent(out)    :: esat, desat
+ class(*), intent(in)     :: temp
+ class(*), intent(out)    :: esat, desat
  integer, intent(out) :: nbad
 
  real    :: tmp, del
  integer :: ind
 
    nbad = 0
-   tmp = temp-tminl
+
+   select type (temp)
+   type is (real(kind=r4_kind))
+     tmp = temp-tminl
+   type is (real(kind=r8_kind))
+     tmp = temp-tminl
+   class default
+     call error_mesg ('sat_vapor_pres_k_mod::lookup_es3_des3_k_0d', 'unsupported kind', FATAL)
+   end select
+
    ind = int(dtinvl*(tmp+tepsl))
    if (ind < 0 .or. ind >= table_siz)  then
      nbad = nbad+1
    else
      del = tmp-dtres*real(ind)
-     esat = TABLE3(ind+1) + &
-            del*(DTABLE3(ind+1) + del*D2TABLE3(ind+1))
-     desat = DTABLE3(ind+1) + 2.*del*D2TABLE3(ind+1)
+     select type (esat)
+     type is (real(kind=r4_kind))
+       esat = TABLE3(ind+1) + del*(DTABLE3(ind+1)+del*D2TABLE3(ind+1))
+     type is (real(kind=r8_kind))
+       esat = TABLE3(ind+1) + del*(DTABLE3(ind+1)+del*D2TABLE3(ind+1))
+     class default
+       call error_mesg ('sat_vapor_pres_k_mod::lookup_es3_des3_k_0d', 'unsupported kind', FATAL)
+     end select
+
+     select type (desat)
+     type is (real(kind=r4_kind))
+       desat = DTABLE3(ind+1) + 2.*del*D2TABLE3(ind+1)
+     type is (real(kind=r8_kind))
+       desat = DTABLE3(ind+1) + 2.*del*D2TABLE3(ind+1)
+     class default
+       call error_mesg ('sat_vapor_pres_k_mod::lookup_es3_des3_k_0d', 'unsupported kind', FATAL)
+     end select
    endif
 
  end subroutine lookup_es3_des3_k_0d
 
 !#######################################################################
 
  subroutine lookup_es3_k_3d(temp, esat, nbad)
- real, intent(in),  dimension(:,:,:)  :: temp
- real, intent(out), dimension(:,:,:)  :: esat
+ class(*), intent(in),  dimension(:,:,:)  :: temp
+ class(*), intent(out), dimension(:,:,:)  :: esat
  integer, intent(out) :: nbad
  real    :: tmp, del
  integer :: ind, i, j, k
+ logical :: valid_types = .false.
+
+   select type (temp)
+   type is (real(kind=r4_kind))
+      select type (esat)
+      type is (real(kind=r4_kind))
+         valid_types = .true.
+      end select
+   type is (real(kind=r8_kind))
+      select type (esat)
+      type is (real(kind=r8_kind))
+         valid_types = .true.
+      end select
+   end select
+   if ( .not. valid_types ) then
+      call error_mesg ('sat_vapor_pres_k_mod::lookup_es3_k_3d', 'unsupported kind', FATAL)
```

Comment:
This error message should be more descriptive.  *"temp and esat types do not match or esat is not one of the supported types of real(kind=4) or real(kind=8)."*